### PR TITLE
Dataset

### PR DESCRIPTION
**Pull request by DavidFM43**
- `train.py`: Refactor
	 - Quitar instalación de `pytorch-segmentation-models` ahora se hace desde el notebook de kaggle.
	 - Nombrar mejor los argumentos cuando se llama el script por CLI.
	 - Añadir las transformaciones de los datasets:
		  - Normalización de los canales de las imagenes.
		  - Antes el dataset retornaba directamente los tensores, ahora eso se maneja con los transforms de torchvision, creo que queda mejor así pero no tengo una razón muy fuerte tampoco.
	 - Random split del dataset en entrenamiento, validación y testeo. La parte de testeo realmente no se usa por lo que es el script de entrenamiento.
- `README.md`: 
	- Añadir sección "Setup" que explica como descargar/procesar los datos, correr el script en entrenamiento en Kaggle y loggear en wandb.
	 - Quitar la sección de TODOs.
- `dataset.py`: Reformat del script, ahora es mas simple.
- `utils.py`: Añadir función para revertir la normalización y obtener imagen original. Mejorar algún docstring.
- `train_kaggle.ipynb`: Notebook para correr el entrenamiento en Kaggle. Remplaza el anterior script `train_kaggle.py`.
- `get_key.py`: Script para configurar la API key de wandb dependiendo si se corre localmente o en Kaggle.
- `prepare.py`: Script para descargar brutos y procesarlos localmente.
- Reorganización de la carpeta `data` en una estructura mas clara y sencilla, se especifica en el `README.md`.
- `requirements.txt`: Actulizar librerias y quitar `torch` y `torchvision` pues la instalación depende del hardware en que se corra.
- `config.yaml`: Antes se llamaba `1.yaml`; Refactor pequeños por estetica.
- `.gitignore`: Añadir el script `key.py` que contiene la llave de wandb.
- `notes/`: Accidentalmente añadí cosas de mas de Obsidian.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -107,7 +110,7 @@
     """
 )
 
-# ij is the number of pixels of class i predicted to belong to class j
+# columns are the predictions and rows are the real labels
 conf_matrix = torch.zeros((7, 7), device=device)
```
Comment (eguar11011) [inline train.py:114]: Debería haber una explicación de como se esta configurando la matriz.

File: utils.py [hunk 2]
```diff
@@ -1,9 +1,17 @@
 import torch
 import torch.nn.functional as F
 from torchvision.utils import draw_segmentation_masks
+import torchvision
 from tqdm import tqdm
 
 
+class UnNormalize(torchvision.transforms.Normalize):
```
Comment (eguar11011) [inline utils.py:8]: Sería bueno dejar un link de porque esto de normalización es bueno. He visto que en algunos repositorios hacen eso,

File: train.py [hunk 3]
```diff
@@ -54,7 +53,7 @@
 target_transform = transforms.Compose(
     [
         transforms.Resize(resize_res),
-        transforms.ToTensor(),
+        transforms.PILToTensor(),
```
Comment (eguar11011) [inline train.py:58]: En el commit creo que debería explicar cual era el error específicamente, porqué no había ningún Issues así que no sé a cual error se refiere.

File: notes/.obsidian/workspace.json [hunk 4]
```diff
@@ -4,16 +4,16 @@
     "type": "split",
     "children": [
       {
-        "id": "3fc41fefca8ecc80",
+        "id": "559065886dc4a506",
```
Comment (eguar11011) [inline notes/.obsidian/workspace.json]: ¿estos cambios de que son? 

Creo que deberíamos colocar en gitignore estos archivos json de obsidian.

File: train.py [hunk 5]
```diff
@@ -1,115 +1,93 @@
...
+        project="ml-experiments",
         config=dict(
             ce_weights=weights,
             epochs=epochs,
             batch_size=batch_size,
             resize_res=resize_res,
             optimizer=type(optimizer).__name__,
-            loss_fn="DiceLoss",
+            # TODO: Automatically put loss fn name
+            loss_fn="dice loss",
```
Comment (eguar11011) [inline train.py]: Aquí falta cambia CE.

File: get_key.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import pkgutil
```
Comment (eguar11011) [inline get_key.py:1]: Creo que dejar un link para buscar sobre esto sería bueno.

File: notebooks/2.0-Data-preprocessing.ipynb [hunk 7]
```diff
@@ -1,102 +0,0 @@
-{
```
Comment (eguar11011) [inline notebooks/2.0-Data-preprocessing.ipynb:1]: ¿Por qué se quito el notebook del preprocesamiento? ya que sería bueno por si alguien quiere revisar como se hizo el preprocesamiento-

File: README.md [hunk 8]
```diff
@@ -1,9 +1,52 @@
...
+- `torchvision >= 2.0.0`
+
+The rest can be installed with pip running:
+```python
+pip install -r requirements.txt
+```
+
+### Running the training script in kaggle
+
+In case you down have a machine with GPU, which is highly needed at this point, you can optionally run the training script on kaggle and user their free GPUs. To do that simply go to the `train_kaggle.ipynb` notebook and click on the botton "Run in kaggle".
```
Comment (eguar11011) [inline README.md]: Ese notebook `train_kaggle.ipynb`no existe.

File: train.py [hunk 9]
```diff
@@ -1,115 +1,96 @@
...
+# crossentropy loss fn weights
+weights = torch.tensor([0.8987, 0.4091, 0.9165, 0.8886, 0.9643, 0.9231, 0.0], device=device)
+loss_fn = nn.CrossEntropyLoss(weight=weights)
 print("CE weights:", weights.tolist())
 
 # log training and data config
 if wandb_log:
-    wandb.login(key="2699e8522063dc2ad0f359c8230e5cc09db3ebd8")
+    wandb.login(key=wandb_key)
     # TODO: Add entity parameter to log to the team account
```
Comment (eguar11011) [inline train.py]: ¿A cual parámetro se refiere?

File: utils.py [hunk 10]
```diff
@@ -1,7 +1,15 @@
 import torch
 import torch.nn.functional as F
 from torchvision.utils import draw_segmentation_masks
-from torchvision.io import read_image
+import torchvision
+from tqdm import tqdm
+
+
+class UnNormalize(torchvision.transforms.Normalize):
+    def __init__(self, mean, std, *args, **kwargs):
```
Comment (eguar11011) [inline utils.py:10]: Falta agregar el docstring

File: train.py (hunk 1 above)
Comment (DavidFM43) [reply train.py:114]: Es la matriz de confusión, a que se refiere con la configuración de la matriz?

File: utils.py (hunk 2 above)
Comment (DavidFM43) [reply utils.py:8]: Hmmm pues creo que es una practica común, en especial cuando usa modelos pre-entrenados, pero no sé si hay un lugar donde expliquen al detalle, usted que ha visto?

File: train.py (hunk 3 above)
Comment (DavidFM43) [reply train.py:58]: Pues el commit creo que se llama "fix scaling bug in labels", osea "corrigiendo bug de escalado en las etiquetas", básicamente es porque la función anterior escalaba los valores pero pues para las etiquetas no se quiere que se modifiquen, entonces ese es el bug. Supongo que podria explicar mejor esto en el commmit, pero tocaria hacer force-push y creo que eso no es muy bueno. Puedo mejor especificar en la descripcion de la PR.

File: notes/.obsidian/workspace.json (hunk 4 above)
Comment (DavidFM43) [reply notes/.obsidian/workspace.json]: Ah esque se me fue unas cosas del obsidian por accidente xd, la verdad no se como manejar esto mejor

File: train.py (hunk 1 above)
Comment (eguar11011) [reply train.py:114]: No dice que esa es la matriz de confusión, entonces yo pensaría que colocar ese comentario de que esa es la matriz de confusión.

File: train.py (hunk 3 above)
Comment (eguar11011) [reply train.py:58]: Si me parece buena idea.

**Review by eguar11011 — CHANGES_REQUESTED** [review]

File: train.py (hunk 5 above)
Comment (DavidFM43) [reply train.py]: Esto esta desactualizado, ese cambio ya se hizó

File: get_key.py (hunk 6 above)
Comment (DavidFM43) [reply get_key.py:1]: Puede ser mas especifico?

File: notebooks/2.0-Data-preprocessing.ipynb (hunk 7 above)
Comment (DavidFM43) [reply notebooks/2.0-Data-preprocessing.ipynb:1]: Porque el ese notebook estaba desactualizado y puede revisar el preprocesamiento en el script `prepare.py`

File: README.md (hunk 8 above)
Comment (DavidFM43) [reply README.md]: Ya existe

File: train.py (hunk 1 above)
Comment (DavidFM43) [reply train.py:114]: Ese comentario ya esta

File: utils.py (hunk 2 above)
Comment (DavidFM43) [reply utils.py:8]: Pues la verdad no lo veo necesario, pero si quiere lo puede poner usted

File: train.py (hunk 3 above)
Comment (DavidFM43) [reply train.py:58]: Ya esta actualizado la descripción de la PR

File: notes/.obsidian/workspace.json (hunk 4 above)
Comment (DavidFM43) [reply notes/.obsidian/workspace.json]: Como asi colocarlos en obsidian?

File: train.py (hunk 9 above)
Comment (DavidFM43) [reply train.py]: Ah si ese comentario ya no es necesario

File: utils.py (hunk 10 above)
Comment (DavidFM43) [reply utils.py:10]: Listo

File: utils.py (hunk 2 above)
Comment (eguar11011) [reply utils.py:8]: Sí, algo así como hacen en este repo: https://github.com/wolny/pytorch-3dunet/blob/master/pytorch3dunet/unet3d/losses.py

File: notes/.obsidian/workspace.json (hunk 4 above)
Comment (eguar11011) [reply notes/.obsidian/workspace.json]: En el gitignore, los archivos json de obsidian.

File: get_key.py (hunk 6 above)
Comment (eguar11011) [reply get_key.py:1]: Si claro. Pensaba en dejar un link de la documentación de eso.

File: notebooks/2.0-Data-preprocessing.ipynb (hunk 7 above)
Comment (eguar11011) [reply notebooks/2.0-Data-preprocessing.ipynb:1]: Aaah listo.

File: get_key.py (hunk 6 above)
Comment (eguar11011) [reply get_key.py:1]: Aunque no lo veo necesario. Podemos dejarlo así.

File: utils.py (hunk 2 above)
Comment (DavidFM43) [reply utils.py:8]: Si pero en ese caso es porque esa idea proviene de un paper, pero creo que normalizar no tiene un autor o una fuente en especifico.

Comment (DavidFM43) [reply utils.py:8]: Yo creo que ponerlo en un todo y si algo se hace despues

File: notes/.obsidian/workspace.json (hunk 4 above)
Comment (DavidFM43) [reply notes/.obsidian/workspace.json]: Listo

File: utils.py (hunk 2 above)
Comment (eguar11011) [reply utils.py:8]: Listo. Yo creo que mejor despues

**Review by eguar11011 — APPROVED** [review]
